### PR TITLE
[DF] Fix ROOT-9977, empty SaveGraph output in some cases

### PR DESCRIPTION
**Pull request by eguiraud**
...and add a test.

### CODE REVIEW

File: tree/dataframe/src/RDFGraphUtils.cxx [hunk 1]
```diff
@@ -61,13 +61,20 @@ std::string GraphCreatorHelper::RepresentGraph(RLoopManager *loopManager)
...
       // Triggers the graph construction. When action->GetGraph() will return, the node will be linked to all the branch
-      leaves.push_back(action->GetGraph());
+      nodes.emplace_back(action->GetGraph());
+   }
+
+   auto edges = loopManager->GetGraphEdges();
+   std::vector<std::shared_ptr<GraphNode>> leaves;
+   for (auto edge : edges) {
+      // Triggers the graph construction. When action->GetGraph() will return, the node will be linked to all the branch
+      nodes.emplace_back(edge->GetGraph());
```
Comment (Axel-Naumann) [inline tree/dataframe/src/RDFGraphUtils.cxx:74]: Or `std::transform` with a prior `nodes.resize(actions.size())` (which also prevents reallocs). But details.

File: tree/dataframe/src/RLoopManager.cxx [hunk 2]
```diff
@@ -688,7 +688,15 @@ std::vector<std::string> RLoopManager::GetFiltersNames()
    return filters;
 }
 
-std::vector<RDFInternal::RActionBase *> RLoopManager::GetAllActions()
+std::vector<RNodeBase *> RLoopManager::GetGraphEdges() const
+{
+   std::vector<RNodeBase *> nodes;
+   nodes.insert(nodes.end(), fBookedFilters.begin(), fBookedFilters.end());
```
Comment (Axel-Naumann) [inline tree/dataframe/src/RLoopManager.cxx:694]: Could `std::vector<RNodeBase *> nodes(fBookedFilters.begin(), fBookedFilters.end());` work?

**Review by Axel-Naumann — APPROVED** [review]

File: tree/dataframe/src/RLoopManager.cxx (hunk 2 above)
Comment (eguiraud) [reply tree/dataframe/src/RLoopManager.cxx]: refactored to remove the extra allocations

File: tree/dataframe/src/RDFGraphUtils.cxx (hunk 1 above)
Comment (eguiraud) [reply tree/dataframe/src/RDFGraphUtils.cxx]: refactored to make it more readable and remove the extra allocations

**Pull request merged**